Doc/library/shutil.rst

```diff
@@ -433,23 +433,42 @@ Directory and files operations
    When no *path* is specified, the results of :func:`os.environ` are used,
    returning either the "PATH" value or a fallback of :attr:`os.defpath`.
 
-   On Windows, the current directory is always prepended to the *path* whether
-   or not you use the default or provide your own, which is the behavior the
-   command shell uses when finding executables.  Additionally, when finding the
-   *cmd* in the *path*, the ``PATHEXT`` environment variable is checked.  For
-   example, if you call ``shutil.which("python")``, :func:`which` will search
-   ``PATHEXT`` to know that it should look for ``python.exe`` within the *path*
-   directories.  For example, on Windows::
+   On Windows, the current directory is prepended to the *path* if
+   the *mode* does not include ``os.X_OK``. When the *mode* does include
+   ``os.X_OK``, the Windows API ``NeedCurrentDirectoryForExePathW`` will be
+   consulted to determine if the current directory should be prepended to
+   *path*.
+
+   Additionally, when finding the *cmd* in the *path*, the ``PATHEXT``
+   environment variable is checked.  For example, if you call
+   ``shutil.which("python")``, :func:`which` will search ``PATHEXT``
+   to know that it should look for ``python.exe`` within the *path*
+   directories. For example, on Windows::
 
       >>> shutil.which("python")
       'C:\\Python33\\python.EXE'
 
+   Similar ``PATHEXT`` logic is also applied when a full path to a *cmd* is
+   given, though without an extension::
```

Comment:
I think we should be able to merge this case into the earlier paragraph, especially since I'm _pretty sure_ this isn't correct? `PATHEXT` is always used, it's just that a full match without anything from `PATHEXT` will match first, so chances are if you provide the extension then it'll find that.

Reply:
Actually this wasn't the case prior to this changeset (it was the main reason I started looking at this issue).. See this example:

```
C:\Users\csm10495\Desktop>python
Python 3.11.0 (main, Oct 24 2022, 18:26:48) [MSC v.1933 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import shutil
>>> shutil.which("C:\\Python3\\python")
>>>
>>> exit()

C:\Users\csm10495\Desktop>cd cpython\

C:\Users\csm10495\Desktop\cpython>python
Running Release|x64 interpreter...
Python 3.12.0a6+ (heads/fix_which_pathext_windows-dirty:5fac84a, Apr  2 2023, 14:41:38) [MSC v.1932 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import shutil
>>> shutil.which("C:\\Python3\\python")
'C:\\Python3\\python.EXE'
>>> exit()
```

So older Python didn't try PATHEXT with a fully given path on Windows.